### PR TITLE
iksemel: Fix encoding of emoji

### PR DESCRIPTION
**Pull request by silkeh**
**Summary**

Add a patch to escape non-ASCII characters.

**Test Plan**

Build this stack and run solbuild with history and verify that polari now includes an encoded emoji:

```
$ gotopkg polari
$ sudo solbuild build -h -d -p local-unstable-x86_64
$ grep Remove pspec_x86_64.xml
* Remove GNOME from networks list &#x1f622; [Jeremy]
```

**Checklist**

- [x] Package was built and tested against unstable

### CODE REVIEW

File: packages/i/iksemel/files/0001-Escape-non-ASCII-characters.patch [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ 	int j = 0;
++	char32_t mb;
++	char buf[13] = {0};
++	mbstate_t state = {0};
+ 
+ 	for (i = 0; i < len; i++) {
+ 		c = src[i];
++
++		int rc = (int)mbrtoc32(&mb, src+i, len - i, &state);
++		if (rc < 0 || rc > 1) {
```
Comment (ikeycode) [inline packages/i/iksemel/files/0001-Escape-non-ASCII-characters.patch:54]: I'm a little worried about this. So, `< 0` is error, but `0` is also returned for NULL input. The check would appear to deal with certain errors, ignore others.. ?

Comment (silkeh) [reply packages/i/iksemel/files/0001-Escape-non-ASCII-characters.patch:57]: Yes. This essentially skips invalid and continuation bytes in the sequence. 

There are three documented error cases for `mbrtoc32`, two of which rely on mult-`char32_t ` sequences (should not be possible within unicode). The third error case happens when an invalid character is encountered, which happens in two situations:

- A continuation byte: should be skipped as it is already encoded.
- An actually invalid sequence: should be skipped in order to produce valid UTF-8 output.

I have added a unit test for both scenarios.

**Review by ikeycode — DISMISSED** [review]
LGTM - worst case we pull from unstable if it goes boom. iksemel is unpleasant to deal with in the best situation.

**Comment by silkeh** [comment]
@ikeycode: I added a patch for non-printable ASCII characters, as that would also break the xml :disappointed:. Biggest risk is me messing up the buffer calculations for some weird character sequences, but I think it should be easy to fix when we encounter something like that.

**Comment by Zaryob** [comment]
See the patch and found it quite clean solution.

**Review by ikeycode — APPROVED** [review]
Necessary evil. Would be nice to get valgrind checks on these soon

**Pull request merged**